bazel/carbon_rules/defs.bzl

```
@@ -23,32 +27,91 @@ def carbon_binary(name, srcs):
         #
         # TODO: Switch to the `prefix_root` based rule similar to linking when
         # the prelude moves there.
-        out = src + ".o"
-        srcs_reordered = [s for s in srcs if s != src] + [src]
-        native.genrule(
-            name = src + ".compile",
-            tools = [
-                "//toolchain/install:prefix_root/bin/carbon",
-                "//toolchain/install:install_data",
-            ],
-            cmd = "$(execpath //toolchain/install:prefix_root/bin/carbon) compile --output=$@ $(SRCS)",
-            srcs = srcs_reordered,
-            outs = [out],
+        out = ctx.actions.declare_file("_objs/{0}/{1}o".format(
+            ctx.label.name,
+            src.short_path.removeprefix(ctx.label.package).removesuffix(src.extension),
+        ))
+        objs.append(out)
+        srcs_reordered = [s for s in ctx.files.srcs if s != src] + [src]
+        ctx.actions.run(
+            outputs = [out],
+            inputs = srcs_reordered,
+            executable = toolchain_driver,
+            tools = depset(toolchain_data),
+            arguments = ["compile", "--output=" + out.path] + [s.path for s in srcs_reordered],
+            mnemonic = "CarbonCompile",
+            progress_message = "Compiling " + src.short_path,
         )
 
     # For now, we assume that the prelude doesn't produce any necessary object
     # code, and don't include the .o files for //core/prelude... in the final
     # linked binary.
     #
     # TODO: This will need to be revisited eventually.
-    objs = [s + ".o" for s in srcs]
-    native.genrule(
-        name = name + ".link",
-        tools = [
-            "//toolchain/install:prefix_root/bin/carbon",
-            "//toolchain/install:install_data",
-        ],
-        cmd = "$(execpath //toolchain/install:prefix_root/bin/carbon) link --output=$@ $(SRCS)",
-        srcs = objs,
-        outs = [name],
+    bin = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.run(
+        outputs = [bin],
+        inputs = objs,
+        executable = toolchain_driver,
+        tools = depset(toolchain_data),
+        arguments = ["link", "--output=" + bin.path] + [o.path for o in objs],
+        mnemonic = "CarbonLink",
+        progress_message = "Linking " + bin.short_path,
+    )
+    return [DefaultInfo(files = depset([bin]))]
+
+carbon_binary_internal = rule(
```

```suggestion
_carbon_binary_internal = rule(
```

(note the call would also need to be changed below)